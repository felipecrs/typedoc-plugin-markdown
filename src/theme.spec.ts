import * as fs from 'fs-extra';
import { UrlMapping } from 'typedoc';

import { setup } from '../test/utils';

describe(`theme`, () => {
  function getExpectedUrls(urlMappings: UrlMapping[]) {
    const expectedUrls = [];
    urlMappings.forEach((urlMapping) => {
      expectedUrls.push(urlMapping.url);
      if (urlMapping.model.children) {
        urlMapping.model.children.forEach((reflection) => {
          expectedUrls.push(reflection.url);
        });
      }
    });
    return expectedUrls;
  }

  let app: any;
  let project: any;
  let theme: any;
  beforeAll(() => {
    ({ app, project } = setup());
    theme = app.renderer.theme;
  });

  describe(`getUrls`, () => {
    test(`should getUrls'`, () => {
      const urlMappings = app.renderer.theme.getUrls(project);
      expect(getExpectedUrls(urlMappings)).toMatchSnapshot();
    });

    test(`should getUrls when readme is defined`, () => {
      const spy = jest.spyOn(app.options, 'getValue').mockReturnValue('./README.md');
      const urlMappings = theme.getUrls(project);
      expect(getExpectedUrls(urlMappings)).toMatchSnapshot();
      spy.mockRestore();
    });
  });

  describe(`output directory`, () => {
    let directoryListingSpy;

    beforeAll(() => {
      directoryListingSpy = jest.spyOn(fs, 'readdirSync');
    });

    test(`should test output directory true with all allowed files and directories`, () => {
      directoryListingSpy.mockReturnValue([
        '.DS_Store',
        'README.md',
        'globals.md',
        'classes',
        'enums',
        'interfaces',
        'media',
        'modules',
      ]);
      expect(theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory true with some files directories`, () => {
      directoryListingSpy.mockReturnValue(['README.md', 'classes', 'media', 'modules']);
      expect(theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory true with just index`, () => {
      directoryListingSpy.mockReturnValue(['README.md']);
      expect(theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory false with unkown index`, () => {
      directoryListingSpy.mockReturnValue(['Unrecognised.md', 'classes', 'enums', 'interfaces', 'media', 'modules']);
      expect(theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false with hidden files`, () => {
      directoryListingSpy.mockReturnValue(['.git', 'classes', 'enums', 'interfaces', 'media', 'modules']);
      expect(theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false without an index`, () => {
      directoryListingSpy.mockReturnValue(['globals.md', 'classes', 'enums', 'interfaces', 'media', 'modules']);
      expect(theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false with unknown folder`, () => {
      directoryListingSpy.mockReturnValue(['README.md', 'folder']);
      expect(theme.isOutputDirectory('/path')).toBeFalsy();
    });
  });

  describe(`output directory`, () => {
    let directoryListingSpy;
    beforeAll(() => {
      directoryListingSpy = jest.spyOn(fs, 'readdirSync');
    });

    test(`should test output directory true with all allowed files and directories`, () => {
      directoryListingSpy.mockReturnValue([
        '.DS_Store',
        'README.md',
        'globals.md',
        'classes',
        'enums',
        'interfaces',
        'media',
        'modules',
      ]);
      expect(theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory true with some files directories`, () => {
      directoryListingSpy.mockReturnValue(['README.md', 'classes', 'media', 'modules']);
      expect(theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory true with just index`, () => {
      directoryListingSpy.mockReturnValue(['README.md']);
      expect(theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory false with unkown index`, () => {
      directoryListingSpy.mockReturnValue(['Unrecognised.md', 'classes', 'enums', 'interfaces', 'media', 'modules']);
      expect(theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false with hidden files`, () => {
      directoryListingSpy.mockReturnValue(['.git', 'classes', 'enums', 'interfaces', 'media', 'modules']);
      expect(theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false without an index`, () => {
      directoryListingSpy.mockReturnValue(['globals.md', 'classes', 'enums', 'interfaces', 'media', 'modules']);
      expect(theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false with unknown folder`, () => {
      directoryListingSpy.mockReturnValue(['README.md', 'folder']);
      expect(theme.isOutputDirectory('/path')).toBeFalsy();
    });
  });
});
