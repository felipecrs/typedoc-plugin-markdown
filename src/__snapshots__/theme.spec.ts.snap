// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownTheme getUrls should getUrls when readme is defined 1`] = `
Array [
  "globals.md",
  "modules/_categories_.md",
  "modules/_classes_.md",
  "modules/_comments_.md",
  "modules/_destrucuting_.md",
  "modules/_functions_.md",
  "modules/_interfaces_.md",
  "modules/_literals_.md",
  "modules/_variables_.md",
  "README.md",
  "modules/_categories_.md",
  "modules/_classes_.md",
  "modules/_comments_.md",
  "modules/_destrucuting_.md",
  "modules/_functions_.md",
  "modules/_interfaces_.md",
  "modules/_literals_.md",
  "modules/_variables_.md",
  "modules/_categories_.md",
  "interfaces/_categories_.categoryinterfacea1.md",
  "interfaces/_categories_.categoryinterfacea2.md",
  "interfaces/_categories_.otherinterfacea.md",
  "modules/_categories_.md#const-categoryvariablea1",
  "modules/_categories_.md#const-categoryvariablea2",
  "modules/_categories_.md#const-categoryvariableb1",
  "modules/_categories_.md#const-othervariablea",
  "modules/_categories_.md#const-othervariableb",
  "modules/_categories_.md#categoryfunctiona1",
  "modules/_categories_.md#categoryfunctiona2",
  "modules/_categories_.md#categoryfunctiona3",
  "modules/_categories_.md#categoryfunctionb2",
  "modules/_categories_.md#otherfunctiona",
  "modules/_categories_.md#otherfunctionb",
  "interfaces/_categories_.categoryinterfacea1.md",
  "interfaces/_categories_.categoryinterfacea1.md#value",
  "interfaces/_categories_.categoryinterfacea2.md",
  "interfaces/_categories_.categoryinterfacea2.md#value",
  "interfaces/_categories_.otherinterfacea.md",
  "interfaces/_categories_.otherinterfacea.md#value",
  "modules/_classes_.md",
  "classes/_classes_.baseclass.md",
  "classes/_classes_.genericclass.md",
  "classes/_classes_.internalclass.md",
  "classes/_classes_.nongenericclass.md",
  "classes/_classes_.subclassa.md",
  "classes/_classes_.subclassb.md",
  "interfaces/_classes_.inameinterface.md",
  "interfaces/_classes_.iprintinterface.md",
  "interfaces/_classes_.iprintnameinterface.md",
  "classes/_classes_.baseclass.md",
  "classes/_classes_.baseclass.md#constructor",
  "classes/_classes_.baseclass.md#private-internalclass",
  "classes/_classes_.baseclass.md#protected-kind",
  "classes/_classes_.baseclass.md#name",
  "classes/_classes_.baseclass.md#static-instance",
  "classes/_classes_.baseclass.md#static-instances",
  "classes/_classes_.baseclass.md#abstract-abstractmethod",
  "classes/_classes_.baseclass.md#arrowfunction",
  "classes/_classes_.baseclass.md#private-checkname",
  "classes/_classes_.baseclass.md#getname",
  "classes/_classes_.baseclass.md#setname",
  "classes/_classes_.baseclass.md#static-catest",
  "classes/_classes_.baseclass.md#static-getinstance",
  "classes/_classes_.baseclass.md#static-getname",
  "classes/_classes_.genericclass.md",
  "classes/_classes_.genericclass.md#constructor",
  "classes/_classes_.genericclass.md#protected-p2",
  "classes/_classes_.genericclass.md#p3",
  "classes/_classes_.genericclass.md#private-p4",
  "classes/_classes_.genericclass.md#readonly-p5",
  "classes/_classes_.genericclass.md#value",
  "classes/_classes_.genericclass.md#getvalue",
  "classes/_classes_.genericclass.md#setvalue",
  "classes/_classes_.internalclass.md",
  "classes/_classes_.internalclass.md#constructor",
  "classes/_classes_.nongenericclass.md",
  "classes/_classes_.nongenericclass.md#constructor",
  "classes/_classes_.nongenericclass.md#protected-p2",
  "classes/_classes_.nongenericclass.md#p3",
  "classes/_classes_.nongenericclass.md#readonly-p5",
  "classes/_classes_.nongenericclass.md#value",
  "classes/_classes_.nongenericclass.md#getvalue",
  "classes/_classes_.nongenericclass.md#setvalue",
  "classes/_classes_.subclassa.md",
  "classes/_classes_.subclassa.md#constructor",
  "classes/_classes_.subclassa.md#protected-kind",
  "classes/_classes_.subclassa.md#name",
  "classes/_classes_.subclassa.md#static-instance",
  "classes/_classes_.subclassa.md#static-instances",
  "classes/_classes_.subclassa.md#nameproperty",
  "classes/_classes_.subclassa.md#readonlynameproperty",
  "classes/_classes_.subclassa.md#writeonlynameproperty",
  "classes/_classes_.subclassa.md#abstractmethod",
  "classes/_classes_.subclassa.md#arrowfunction",
  "classes/_classes_.subclassa.md#getname",
  "classes/_classes_.subclassa.md#print",
  "classes/_classes_.subclassa.md#printname",
  "classes/_classes_.subclassa.md#setname",
  "classes/_classes_.subclassa.md#static-catest",
  "classes/_classes_.subclassa.md#static-getinstance",
  "classes/_classes_.subclassa.md#static-getname",
  "classes/_classes_.subclassb.md",
  "classes/_classes_.subclassb.md#constructor",
  "classes/_classes_.subclassb.md#protected-kind",
  "classes/_classes_.subclassb.md#name",
  "classes/_classes_.subclassb.md#static-instance",
  "classes/_classes_.subclassb.md#static-instances",
  "classes/_classes_.subclassb.md#abstractmethod",
  "classes/_classes_.subclassb.md#arrowfunction",
  "classes/_classes_.subclassb.md#dosomething",
  "classes/_classes_.subclassb.md#getname",
  "classes/_classes_.subclassb.md#setname",
  "classes/_classes_.subclassb.md#static-catest",
  "classes/_classes_.subclassb.md#static-getinstance",
  "classes/_classes_.subclassb.md#static-getname",
  "interfaces/_classes_.inameinterface.md",
  "interfaces/_classes_.inameinterface.md#name",
  "interfaces/_classes_.inameinterface.md#getname",
  "interfaces/_classes_.iprintinterface.md",
  "interfaces/_classes_.iprintinterface.md#print",
  "interfaces/_classes_.iprintnameinterface.md",
  "interfaces/_classes_.iprintnameinterface.md#name",
  "interfaces/_classes_.iprintnameinterface.md#getname",
  "interfaces/_classes_.iprintnameinterface.md#print",
  "interfaces/_classes_.iprintnameinterface.md#printname",
  "modules/_comments_.md",
  "modules/_comments_.md#const-commentswithfencedblock",
  "modules/_comments_.md#const-commentswithincludes",
  "modules/_comments_.md#const-commentswithsymbollinks",
  "modules/_comments_.md#const-commentswithtags",
  "modules/_comments_.md#commentsinreturn",
  "modules/_comments_.md#functionwithdoclink",
  "modules/_destrucuting_.md",
  "interfaces/_destrucuting_.i.md",
  "modules/_destrucuting_.md#destructarraya",
  "modules/_destrucuting_.md#destructarrayb",
  "modules/_destrucuting_.md#destructarrayc",
  "modules/_destrucuting_.md#destructarraywithignoresa",
  "modules/_destrucuting_.md#destructarraywithignoresrest",
  "modules/_destrucuting_.md#destructarraywithrest",
  "modules/_destrucuting_.md#destructarraywithresta",
  "modules/_destrucuting_.md#destructarraywithrestb",
  "modules/_destrucuting_.md#destructobjecta",
  "modules/_destrucuting_.md#destructobjectb",
  "modules/_destrucuting_.md#destructobjectc",
  "modules/_destrucuting_.md#drawtext",
  "interfaces/_destrucuting_.i.md",
  "interfaces/_destrucuting_.i.md#name",
  "modules/_functions_.md",
  "modules/_functions_.md#buildname",
  "modules/_functions_.md#exportedfunction",
  "modules/_functions_.md#functionwithdefaults",
  "modules/_functions_.md#functionwithoptionalvalue",
  "modules/_functions_.md#functionwithparameters",
  "modules/_functions_.md#functionwithrest",
  "modules/_functions_.md#internalfunction",
  "modules/_functions_.md#const-variablefunction",
  "modules/_interfaces_.md",
  "classes/_interfaces_.control.md",
  "interfaces/_interfaces_.indexableinterface.md",
  "interfaces/_interfaces_.labeledvalue.md",
  "interfaces/_interfaces_.mixedproperties.md",
  "interfaces/_interfaces_.penstroke.md",
  "interfaces/_interfaces_.selectablecontrol.md",
  "interfaces/_interfaces_.shape.md",
  "interfaces/_interfaces_.square.md",
  "modules/_interfaces_.md#printlabel",
  "classes/_interfaces_.control.md",
  "classes/_interfaces_.control.md#state",
  "interfaces/_interfaces_.indexableinterface.md",
  "interfaces/_interfaces_.labeledvalue.md",
  "interfaces/_interfaces_.labeledvalue.md#label",
  "interfaces/_interfaces_.mixedproperties.md",
  "interfaces/_interfaces_.mixedproperties.md#optional-color",
  "interfaces/_interfaces_.mixedproperties.md#control",
  "interfaces/_interfaces_.mixedproperties.md#mixedobject",
  "interfaces/_interfaces_.mixedproperties.md#nestedmixedobject",
  "interfaces/_interfaces_.mixedproperties.md#shape",
  "interfaces/_interfaces_.mixedproperties.md#width",
  "interfaces/_interfaces_.mixedproperties.md#optional-readonly-x",
  "interfaces/_interfaces_.mixedproperties.md#readonly-y",
  "interfaces/_interfaces_.mixedproperties.md#select",
  "interfaces/_interfaces_.penstroke.md",
  "interfaces/_interfaces_.penstroke.md#penwidth",
  "interfaces/_interfaces_.selectablecontrol.md",
  "interfaces/_interfaces_.selectablecontrol.md#state",
  "interfaces/_interfaces_.selectablecontrol.md#select",
  "interfaces/_interfaces_.shape.md",
  "interfaces/_interfaces_.shape.md#color",
  "interfaces/_interfaces_.square.md",
  "interfaces/_interfaces_.square.md#color",
  "interfaces/_interfaces_.square.md#penwidth",
  "interfaces/_interfaces_.square.md#sidelength",
  "modules/_literals_.md",
  "modules/_literals_.md#let-typeliteral",
  "modules/_literals_.md#const-objectliteral",
  "modules/_variables_.md",
  "modules/_variables_.md#const-color",
  "modules/_variables_.md#const-decimal",
  "modules/_variables_.md#const-isdone",
  "modules/_variables_.md#const-list",
  "modules/_variables_.md#const-n",
  "modules/_variables_.md#const-notsure",
  "modules/_variables_.md#let-u",
  "modules/_variables_.md#let-x",
]
`;

exports[`MarkdownTheme getUrls should getUrls' 1`] = `
Array [
  "README.md",
  "modules/_categories_.md",
  "modules/_classes_.md",
  "modules/_comments_.md",
  "modules/_destrucuting_.md",
  "modules/_functions_.md",
  "modules/_interfaces_.md",
  "modules/_literals_.md",
  "modules/_variables_.md",
  "modules/_categories_.md",
  "interfaces/_categories_.categoryinterfacea1.md",
  "interfaces/_categories_.categoryinterfacea2.md",
  "interfaces/_categories_.otherinterfacea.md",
  "modules/_categories_.md#const-categoryvariablea1",
  "modules/_categories_.md#const-categoryvariablea2",
  "modules/_categories_.md#const-categoryvariableb1",
  "modules/_categories_.md#const-othervariablea",
  "modules/_categories_.md#const-othervariableb",
  "modules/_categories_.md#categoryfunctiona1",
  "modules/_categories_.md#categoryfunctiona2",
  "modules/_categories_.md#categoryfunctiona3",
  "modules/_categories_.md#categoryfunctionb2",
  "modules/_categories_.md#otherfunctiona",
  "modules/_categories_.md#otherfunctionb",
  "interfaces/_categories_.categoryinterfacea1.md",
  "interfaces/_categories_.categoryinterfacea1.md#value",
  "interfaces/_categories_.categoryinterfacea2.md",
  "interfaces/_categories_.categoryinterfacea2.md#value",
  "interfaces/_categories_.otherinterfacea.md",
  "interfaces/_categories_.otherinterfacea.md#value",
  "modules/_classes_.md",
  "classes/_classes_.baseclass.md",
  "classes/_classes_.genericclass.md",
  "classes/_classes_.internalclass.md",
  "classes/_classes_.nongenericclass.md",
  "classes/_classes_.subclassa.md",
  "classes/_classes_.subclassb.md",
  "interfaces/_classes_.inameinterface.md",
  "interfaces/_classes_.iprintinterface.md",
  "interfaces/_classes_.iprintnameinterface.md",
  "classes/_classes_.baseclass.md",
  "classes/_classes_.baseclass.md#constructor",
  "classes/_classes_.baseclass.md#private-internalclass",
  "classes/_classes_.baseclass.md#protected-kind",
  "classes/_classes_.baseclass.md#name",
  "classes/_classes_.baseclass.md#static-instance",
  "classes/_classes_.baseclass.md#static-instances",
  "classes/_classes_.baseclass.md#abstract-abstractmethod",
  "classes/_classes_.baseclass.md#arrowfunction",
  "classes/_classes_.baseclass.md#private-checkname",
  "classes/_classes_.baseclass.md#getname",
  "classes/_classes_.baseclass.md#setname",
  "classes/_classes_.baseclass.md#static-catest",
  "classes/_classes_.baseclass.md#static-getinstance",
  "classes/_classes_.baseclass.md#static-getname",
  "classes/_classes_.genericclass.md",
  "classes/_classes_.genericclass.md#constructor",
  "classes/_classes_.genericclass.md#protected-p2",
  "classes/_classes_.genericclass.md#p3",
  "classes/_classes_.genericclass.md#private-p4",
  "classes/_classes_.genericclass.md#readonly-p5",
  "classes/_classes_.genericclass.md#value",
  "classes/_classes_.genericclass.md#getvalue",
  "classes/_classes_.genericclass.md#setvalue",
  "classes/_classes_.internalclass.md",
  "classes/_classes_.internalclass.md#constructor",
  "classes/_classes_.nongenericclass.md",
  "classes/_classes_.nongenericclass.md#constructor",
  "classes/_classes_.nongenericclass.md#protected-p2",
  "classes/_classes_.nongenericclass.md#p3",
  "classes/_classes_.nongenericclass.md#readonly-p5",
  "classes/_classes_.nongenericclass.md#value",
  "classes/_classes_.nongenericclass.md#getvalue",
  "classes/_classes_.nongenericclass.md#setvalue",
  "classes/_classes_.subclassa.md",
  "classes/_classes_.subclassa.md#constructor",
  "classes/_classes_.subclassa.md#protected-kind",
  "classes/_classes_.subclassa.md#name",
  "classes/_classes_.subclassa.md#static-instance",
  "classes/_classes_.subclassa.md#static-instances",
  "classes/_classes_.subclassa.md#nameproperty",
  "classes/_classes_.subclassa.md#readonlynameproperty",
  "classes/_classes_.subclassa.md#writeonlynameproperty",
  "classes/_classes_.subclassa.md#abstractmethod",
  "classes/_classes_.subclassa.md#arrowfunction",
  "classes/_classes_.subclassa.md#getname",
  "classes/_classes_.subclassa.md#print",
  "classes/_classes_.subclassa.md#printname",
  "classes/_classes_.subclassa.md#setname",
  "classes/_classes_.subclassa.md#static-catest",
  "classes/_classes_.subclassa.md#static-getinstance",
  "classes/_classes_.subclassa.md#static-getname",
  "classes/_classes_.subclassb.md",
  "classes/_classes_.subclassb.md#constructor",
  "classes/_classes_.subclassb.md#protected-kind",
  "classes/_classes_.subclassb.md#name",
  "classes/_classes_.subclassb.md#static-instance",
  "classes/_classes_.subclassb.md#static-instances",
  "classes/_classes_.subclassb.md#abstractmethod",
  "classes/_classes_.subclassb.md#arrowfunction",
  "classes/_classes_.subclassb.md#dosomething",
  "classes/_classes_.subclassb.md#getname",
  "classes/_classes_.subclassb.md#setname",
  "classes/_classes_.subclassb.md#static-catest",
  "classes/_classes_.subclassb.md#static-getinstance",
  "classes/_classes_.subclassb.md#static-getname",
  "interfaces/_classes_.inameinterface.md",
  "interfaces/_classes_.inameinterface.md#name",
  "interfaces/_classes_.inameinterface.md#getname",
  "interfaces/_classes_.iprintinterface.md",
  "interfaces/_classes_.iprintinterface.md#print",
  "interfaces/_classes_.iprintnameinterface.md",
  "interfaces/_classes_.iprintnameinterface.md#name",
  "interfaces/_classes_.iprintnameinterface.md#getname",
  "interfaces/_classes_.iprintnameinterface.md#print",
  "interfaces/_classes_.iprintnameinterface.md#printname",
  "modules/_comments_.md",
  "modules/_comments_.md#const-commentswithfencedblock",
  "modules/_comments_.md#const-commentswithincludes",
  "modules/_comments_.md#const-commentswithsymbollinks",
  "modules/_comments_.md#const-commentswithtags",
  "modules/_comments_.md#commentsinreturn",
  "modules/_comments_.md#functionwithdoclink",
  "modules/_destrucuting_.md",
  "interfaces/_destrucuting_.i.md",
  "modules/_destrucuting_.md#destructarraya",
  "modules/_destrucuting_.md#destructarrayb",
  "modules/_destrucuting_.md#destructarrayc",
  "modules/_destrucuting_.md#destructarraywithignoresa",
  "modules/_destrucuting_.md#destructarraywithignoresrest",
  "modules/_destrucuting_.md#destructarraywithrest",
  "modules/_destrucuting_.md#destructarraywithresta",
  "modules/_destrucuting_.md#destructarraywithrestb",
  "modules/_destrucuting_.md#destructobjecta",
  "modules/_destrucuting_.md#destructobjectb",
  "modules/_destrucuting_.md#destructobjectc",
  "modules/_destrucuting_.md#drawtext",
  "interfaces/_destrucuting_.i.md",
  "interfaces/_destrucuting_.i.md#name",
  "modules/_functions_.md",
  "modules/_functions_.md#buildname",
  "modules/_functions_.md#exportedfunction",
  "modules/_functions_.md#functionwithdefaults",
  "modules/_functions_.md#functionwithoptionalvalue",
  "modules/_functions_.md#functionwithparameters",
  "modules/_functions_.md#functionwithrest",
  "modules/_functions_.md#internalfunction",
  "modules/_functions_.md#const-variablefunction",
  "modules/_interfaces_.md",
  "classes/_interfaces_.control.md",
  "interfaces/_interfaces_.indexableinterface.md",
  "interfaces/_interfaces_.labeledvalue.md",
  "interfaces/_interfaces_.mixedproperties.md",
  "interfaces/_interfaces_.penstroke.md",
  "interfaces/_interfaces_.selectablecontrol.md",
  "interfaces/_interfaces_.shape.md",
  "interfaces/_interfaces_.square.md",
  "modules/_interfaces_.md#printlabel",
  "classes/_interfaces_.control.md",
  "classes/_interfaces_.control.md#state",
  "interfaces/_interfaces_.indexableinterface.md",
  "interfaces/_interfaces_.labeledvalue.md",
  "interfaces/_interfaces_.labeledvalue.md#label",
  "interfaces/_interfaces_.mixedproperties.md",
  "interfaces/_interfaces_.mixedproperties.md#optional-color",
  "interfaces/_interfaces_.mixedproperties.md#control",
  "interfaces/_interfaces_.mixedproperties.md#mixedobject",
  "interfaces/_interfaces_.mixedproperties.md#nestedmixedobject",
  "interfaces/_interfaces_.mixedproperties.md#shape",
  "interfaces/_interfaces_.mixedproperties.md#width",
  "interfaces/_interfaces_.mixedproperties.md#optional-readonly-x",
  "interfaces/_interfaces_.mixedproperties.md#readonly-y",
  "interfaces/_interfaces_.mixedproperties.md#select",
  "interfaces/_interfaces_.penstroke.md",
  "interfaces/_interfaces_.penstroke.md#penwidth",
  "interfaces/_interfaces_.selectablecontrol.md",
  "interfaces/_interfaces_.selectablecontrol.md#state",
  "interfaces/_interfaces_.selectablecontrol.md#select",
  "interfaces/_interfaces_.shape.md",
  "interfaces/_interfaces_.shape.md#color",
  "interfaces/_interfaces_.square.md",
  "interfaces/_interfaces_.square.md#color",
  "interfaces/_interfaces_.square.md#penwidth",
  "interfaces/_interfaces_.square.md#sidelength",
  "modules/_literals_.md",
  "modules/_literals_.md#let-typeliteral",
  "modules/_literals_.md#const-objectliteral",
  "modules/_variables_.md",
  "modules/_variables_.md#const-color",
  "modules/_variables_.md#const-decimal",
  "modules/_variables_.md#const-isdone",
  "modules/_variables_.md#const-list",
  "modules/_variables_.md#const-n",
  "modules/_variables_.md#const-notsure",
  "modules/_variables_.md#let-u",
  "modules/_variables_.md#let-x",
]
`;
